python/lsst/ts/envsensors/device/base_device.py

```diff
@@ -0,0 +1,142 @@
+
+    Parameters:
+    -----------
+    name: `str`
+        The name of the device.
+    device_id: `str`
+        The hardware device ID to connect to. This can be a physical ID (e.g.
+        /dev/ttyUSB0), a serial port (e.g. serial_ch_1) or any other ID used by
+        the specific device.
+    sensor: `BaseSensor`
+        The sensor that produces the telemetry.
+    callback_func : `Callable`
+        Callback function to receive the telemetry.
+    log: `logging.Logger`
+        The logger to create a child logger for.
+    """
+
+    @abstractmethod
+    def __init__(
+        self,
+        name: str,
+        device_id: str,
+        sensor: BaseSensor,
+        callback_func: Callable,
+        log: logging.Logger,
+    ) -> None:
+        self.name: str = name
+        self._device_id: str = device_id
+        self._sensor: BaseSensor = sensor
+        self._callback_func: Callable = callback_func
+        self._telemetry_loop: asyncio.Future = create_done_future()
+        self._log = log.getChild(type(self).__name__)
+
+    @abstractmethod
+    def init(self) -> None:
+        """Initialize the Sensor Device."""
+        pass
+
+    @abstractmethod
+    async def open(self) -> None:
+        """Open the Sensor Device."""
+        pass
+
+    async def start(self) -> None:
+        """Start the sensor read loop."""
+        self._log.debug(f"Starting read loop for {self.name!r} sensor.")
+        self._telemetry_loop = asyncio.create_task(self._run())
+
+    async def _run(self) -> None:
+        """Run sensor read loop.
+
+        If enabled, loop and read the sensor and pass result to callback_func.
+        """
+        self._log.debug("Starting sensor.")
+        while not self._telemetry_loop.done():
+            self._log.debug("Reading data.")
+            tm: float = time.time()
+            response: int = ResponseCode.OK
+            line: str = await self.readline()
+            sensor_telemetry: List[float] = await self._sensor.extract_telemetry(
+                line=line
+            )
+            output: List[Any] = [
+                self.name,
+                tm,
+                response,
+            ] + sensor_telemetry  # type: ignore
+            reply = {
+                Key.TELEMETRY: output,
+            }
+            self._log.info(f"Returning {reply}")
+            await self._callback_func(reply)
+
+    @abstractmethod
+    async def readline(self) -> str:
+        """Read a line of telemetry from the Device.
+
+        Returns
+        -------
+        output: `str`
+            A line of telemetry.
+        """
+        pass
+
+    async def stop(self) -> None:
+        """Terminate the sensor read loop."""
+        self._log.debug(f"Stopping read loop for {self.name!r} sensor.")
+        self._telemetry_loop.cancel()
+        self._telemetry_loop = create_done_future()
+
+    @abstractmethod
+    async def close(self) -> None:
+        """Close the Sensor Device."""
+        pass
```

Comment:
Please consider suggest tracking whether the device is open. This add a bit of complexity but offers some advantages:
- readline can be safer: reject calls unless open
- open can be safer: reject attempts to open if already open
- close can be idempotent (if closed then silently do nothing)

Here is an example, where `open` and `close` are now concrete methods that call abstract methods `basic_open` and `basic_close` to do the work. For example:
```
def __init__(...):
    self.is_open = False

async def open(self) -> None:
    if self.is_open:
        raise RuntimeError("Already open")
    await self.basic_open()
    self.is_open = True

async def close(self) -> None:
    if not self.is_open:
        return
    self.is_open = False
    await self.basic_close()
```
You may also want to think about how to handle exceptions in basic_open and basic_close. The code above assumes that basic_open leaves the connection closed if it fails (I think this is a reasonable requirement for concrete subclasses). If basic_close fails then you have done all you can, but you may end up unable to open the connect again.

You can also make state a DeviceState enum value with OPEN, OPENING, CLOSING, CLOSED values, if you prefer. But if you do that please be careful not to get stuck in an unusable intermediate state if `basic_open` or `basic_close` raise an exception.

Reply:
I adopted the generic open and close functions. Exception handling in basic_open and basic_close will come at a later stage when I can test the code with the specific devices.